File: touchflow/external_kb.py
"""Обнаружение подключённых физических клавиатур."""

from __future__ import annotations

import logging
import re
from pathlib import Path

log = logging.getLogger(__name__)

INPUT_DEVICES = Path("/proc/bus/input/devices")
KEYBOARD_CAP = re.compile(r"\bEV_KEY\b")


def _parse_devices(text: str) -> list[dict[str, str]]:
    blocks = text.strip().split("\n\n")
    devices: list[dict[str, str]] = []
    for block in blocks:
        info: dict[str, str] = {}
        for line in block.splitlines():
            if line.startswith("I:"):
                m = re.search(r"bus=(\S+)", line)
                if m:
                    info["bus"] = m.group(1)
            elif line.startswith("N:"):
                info["name"] = line.split("Name=", 1)[-1].strip().strip('"')
            elif line.startswith("H:"):
                info["handlers"] = line.split("Handlers=", 1)[-1].strip()
            elif line.startswith("B:") and "EV_KEY" in line:
                info["has_keys"] = "1"
        if info.get("has_keys"):
            devices.append(info)
    return devices


def list_keyboards() -> list[dict[str, str]]:
    if not INPUT_DEVICES.exists():
        return []
    text = INPUT_DEVICES.read_text(encoding="utf-8", errors="replace")
    return _parse_devices(text)


def has_external_keyboard() -> bool:
    """True если обнаружена хотя бы одна физическая клавиатура (не виртуальная)."""
    for dev in list_keyboards():
        name = dev.get("name", "").lower()
        bus = dev.get("bus", "")
        handlers = dev.get("handlers", "")

        # Пропускаем виртуальные устройства и наши собственные
        skip_patterns = (
            "touchflow",
            "virtual",
            "uinput",
            "dummy",
            "power button",
            "sleep button",
            "video bus",
            "gpio",
        )
        if any(p in name for p in skip_patterns):
            continue

        # evdev handler eventN
        if "event" not in handlers:
            continue

        # USB / Bluetooth / I2C HID клавиатуры
        if bus in ("0003", "0005", "0018") or "kbd" in handlers:
            log.debug("External keyboard: %s (%s)", name, bus)
            return True

        # Встроенные клавиатуры ноутбуков (i8042)
        if bus == "0011" and "kbd" in handlers:
            log.debug("Built-in keyboard: %s", name)
            return True

    return False


class ExternalKeyboardMonitor:
    """Периодический опрос /proc/bus/input/devices (надёжно на всех дистрибутивах)."""

    def __init__(self, on_change=None):
        self._on_change = on_change
        self._connected = has_external_keyboard()

    @property
    def connected(self) -> bool:
        return self._connected

    def poll(self) -> bool:
        current = has_external_keyboard()
        if current != self._connected:
            self._connected = current
            if self._on_change:
                self._on_change(current)
        return current

File: experimental/touchflow-cpp/src/external_kb.cpp
#include "external_kb.hpp"

#include <fstream>
#include <string>
#include <algorithm>
#include <cctype>

namespace touchflow {

static std::string lower(const std::string& s) {
    std::string r = s;
    std::transform(r.begin(), r.end(), r.begin(), [](unsigned char c) { return std::tolower(c); });
    return r;
}

static bool should_skip(const std::string& name) {
    static const char* patterns[] = {
        "touchflow", "virtual", "uinput", "dummy", "power button",
        "sleep button", "video bus", "gpio", nullptr
    };
    auto n = lower(name);
    for (const char** p = patterns; *p; ++p)
        if (n.find(*p) != std::string::npos) return true;
    return false;
}

ExternalKeyboardMonitor::ExternalKeyboardMonitor(Callback cb, void* data)
    : callback_(cb), data_(data) {
    connected_ = detect();
}

bool ExternalKeyboardMonitor::detect() const {
    std::ifstream in("/proc/bus/input/devices");
    if (!in) return false;
    std::string block, line;
    bool has_keys = false;
    std::string name, bus;
    auto flush = [&]() {
        if (has_keys && !name.empty() && !should_skip(name)) {
            if (bus.find("0019") == std::string::npos) return true;
        }
        has_keys = false;
        name.clear();
        bus.clear();
        return false;
    };
    while (std::getline(in, line)) {
        if (line.empty()) {
            if (flush()) return true;
            continue;
        }
        if (line.rfind("N:", 0) == 0) {
            auto pos = line.find("Name=");
            if (pos != std::string::npos)
                name = line.substr(pos + 5);
            while (!name.empty() && (name.front() == '"' || name.front() == '\'')) name.erase(name.begin());
            while (!name.empty() && (name.back() == '"' || name.back() == '\'')) name.pop_back();
        } else if (line.rfind("I:", 0) == 0) {
            auto pos = line.find("bus=");
            if (pos != std::string::npos) bus = line.substr(pos + 4);
        } else if (line.rfind("B:", 0) == 0 && line.find("EV_KEY") != std::string::npos) {
            has_keys = true;
        }
    }
    return flush();
}

void ExternalKeyboardMonitor::poll() {
    bool now = detect();
    if (now != connected_) {
        connected_ = now;
        if (callback_) callback_(connected_, data_);
    }
}

bool ExternalKeyboardMonitor::connected() const { return connected_; }

}  // namespace touchflow
